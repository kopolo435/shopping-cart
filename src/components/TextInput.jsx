import React from "react";
import PropTypes from "prop-types";

function TextInput({
  type,
  label,
  name,
  updateFormErrors,
  updateFormValues,
  validation,
}) {
  const [value, setValue] = React.useState("");
  const [error, setError] = React.useState(false);

  function handleInputChange(newValue) {
    setValue(newValue);
    updateFormValues(newValue);
    if (!validation(newValue)) {
      setError(true);
      updateFormErrors((current) => {
        const tempMap = new Map([...current]);
        tempMap.set(name, true);
        return tempMap;
      });
    }
  }

  return !error ? (
    <div className="textInputContainer">
      <label htmlFor="nombre">
        {label}
        *
        <input
          type={type}
          id={name}
          value={value}
          name={name}
          required
          onChange={(e) => handleInputChange(e.target.value)}
        />
      </label>
    </div>
  ) : (
    <div className="textInputContainer">
      <label htmlFor="nombre">
        {label}
        *
        <input
          type={type}
          id={name}
          value={value}
          name={name}
          aria-describedby="nombre-error"
          aria-invalid="true"
          required
          onChange={(e) => handleInputChange(e.target.value)}
        />
      </label>
      <p id="nombre-error" role="alert">
        El nombre no es del tama√±o correcto
      </p>
    </div>
  );
}

TextInput.propTypes = {
  type: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  updateFormErrors: PropTypes.func.isRequired,
  updateFormValues: PropTypes.func.isRequired,
  validation: PropTypes.func.isRequired,
};

export default TextInput;
