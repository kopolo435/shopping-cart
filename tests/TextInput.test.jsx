import { getByTestId, render, screen } from "@testing-library/react";
import { expect, it, vi } from "vitest";
import React from "react";
import userEvent from "@testing-library/user-event";
import { node } from "prop-types";
import TextInput from "../src/components/TextInput";

it("renders input right", () => {
  const { container } = render(
    <TextInput
      type="text"
      label="Nombre"
      name="nombre"
      updateFormErrors={() => {}}
      updateFormValues={() => {}}
      validation={() => {}}
    />
  );

  expect(container).toMatchSnapshot();
});

it("Calls updateFormError and UpdateFormValues on an error", async () => {
  const mock1Fn = vi.fn(() => {});
  const mock2Fn = vi.fn(() => {});
  const user = userEvent.setup();
  render(
    <TextInput
      type="text"
      label="Nombre"
      name="nombre"
      updateFormErrors={mock1Fn}
      updateFormValues={mock2Fn}
      validation={() => "Hay un error grave"}
    />
  );

  const input = screen.getByRole("textbox", { name: "Nombre*" });

  await user.type(input, "1231");

  expect(mock1Fn).toHaveBeenCalled();
  expect(mock2Fn).toHaveBeenCalled();
});

it("Calls updateFormValues on input change", async () => {
  const mockFn = vi.fn(() => {});
  const user = userEvent.setup();
  render(
    <TextInput
      type="text"
      label="Nombre"
      name="nombre"
      updateFormErrors={() => {}}
      updateFormValues={mockFn}
      validation={() => ""}
    />
  );

  const input = screen.getByRole("textbox", { name: "Nombre*" });

  await user.type(input, "Jose");

  expect(mockFn).toBeCalled();
});

it("Shows alert on an error", async () => {
  const user = userEvent.setup();
  render(
    <TextInput
      type="text"
      label="Nombre"
      name="nombre"
      updateFormErrors={() => {}}
      updateFormValues={() => {}}
      validation={() => "Hay un error grave"}
    />
  );

  const input = screen.getByRole("textbox", { name: "Nombre*" });

  await user.type(input, "1231");

  expect(screen.getByRole("alert").textContent.length > 0).toBe(true);
});
