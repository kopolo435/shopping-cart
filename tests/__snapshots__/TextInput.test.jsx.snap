// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders input right 1`] = `
<div>
  <div
    class="textInputContainer"
  >
    <label
      for="nombre"
    >
      Nombre
      *
      <input
        id="nombre"
        name="nombre"
        required=""
        type="text"
        value=""
      />
    </label>
  </div>
</div>
`;
